File: sticker.php
<?php
// sticker.php
// Sticker Maker на PHP (использует расширение imagick)

if (php_sapi_name() !== 'cli') {
    die("Это консольное приложение.\n");
}

$options = getopt('o:c:t:vh', ['output:', 'color:', 'threshold:', 'verbose', 'help']);
$args = array_slice($argv, 1);
$input = null;
foreach ($args as $arg) {
    if (!str_starts_with($arg, '-')) {
        $input = $arg;
        break;
    }
}

if (isset($options['h']) || isset($options['help']) || !$input) {
    echo "Использование: php sticker.php [опции] <входной_файл>\n";
    echo "  -o, --output <файл>   Выходной файл (PNG)\n";
    echo "  -c, --color <HEX>     Цвет фона\n";
    echo "  -t, --threshold <N>   Порог (0-100)\n";
    echo "  -v, --verbose         Подробный вывод\n";
    echo "  -h, --help            Справка\n";
    exit(0);
}

if (!file_exists($input)) {
    fwrite(STDERR, "Ошибка: файл $input не найден\n");
    exit(1);
}

$output = $options['o'] ?? $options['output'] ?? null;
if (!$output) {
    $output = preg_replace('/\.[^.]+$/', '', $input) . '_sticker.png';
}

$colorHex = $options['c'] ?? $options['color'] ?? null;
$threshold = isset($options['t']) ? (int)$options['t'] : (isset($options['threshold']) ? (int)$options['threshold'] : 30);
$verbose = isset($options['v']) || isset($options['verbose']);

try {
    $img = new Imagick($input);
    $w = $img->getImageWidth();
    $h = $img->getImageHeight();

    // Определяем цвет фона
    if ($colorHex) {
        $bgColor = new ImagickPixel($colorHex);
    } else {
        // Автоопределение: средний цвет краёв
        $pixels = [];
        // Верхняя и нижняя строки
        for ($x = 0; $x < $w; $x++) {
            $pixels[] = $img->getImagePixelColor($x, 0);
            $pixels[] = $img->getImagePixelColor($x, $h - 1);
        }
        for ($y = 0; $y < $h; $y++) {
            $pixels[] = $img->getImagePixelColor(0, $y);
            $pixels[] = $img->getImagePixelColor($w - 1, $y);
        }
        // Усредняем
        $sr = $sg = $sb = 0;
        foreach ($pixels as $p) {
            $color = $p->getColor();
            $sr += $color['r'];
            $sg += $color['g'];
            $sb += $color['b'];
        }
        $n = count($pixels);
        if ($n > 0) {
            $bgColor = new ImagickPixel(sprintf("rgb(%d,%d,%d)", $sr/$n, $sg/$n, $sb/$n));
        } else {
            $bgColor = new ImagickPixel('white');
        }
    }

    if ($verbose) {
        $c = $bgColor->getColor();
        echo "Цвет фона: RGB({$c['r']},{$c['g']},{$c['b']})\n";
    }

    // Создаём изображение с альфа-каналом
    $img->setImageMatte(true);
    $maxDist = sqrt(3 * 255 * 255);
    $thresh = ($threshold / 100.0) * $maxDist;

    // Проходим по пикселям
    $bgR = $bgColor->getColor()['r'];
    $bgG = $bgColor->getColor()['g'];
    $bgB = $bgColor->getColor()['b'];

    for ($y = 0; $y < $h; $y++) {
        for ($x = 0; $x < $w; $x++) {
            $px = $img->getImagePixelColor($x, $y);
            $c = $px->getColor();
            $dr = $c['r'] - $bgR;
            $dg = $c['g'] - $bgG;
            $db = $c['b'] - $bgB;
            $dist = sqrt($dr*$dr + $dg*$dg + $db*$db);
            if ($dist <= $thresh) {
                $img->setImagePixelColor($x, $y, new ImagickPixel('rgba(0,0,0,0)'));
            } else {
                $img->setImagePixelColor($x, $y, new ImagickPixel("rgba({$c['r']},{$c['g']},{$c['b']},1)"));
            }
        }
    }

    $img->setImageFormat('png');
    $img->writeImage($output);
    $img->destroy();

    if ($verbose) {
        echo "Фон удалён. Результат сохранён в $output\n";
    }
} catch (Exception $e) {
    fwrite(STDERR, "Ошибка: " . $e->getMessage() . "\n");
    exit(1);
}

File: sticker.rb
# sticker.rb
# Sticker Maker на Ruby (использует mini_magick)

require 'mini_magick'
require 'optparse'

options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby sticker.rb [опции] <входной_файл>"
  opts.on('-o', '--output FILE', 'Выходной файл (PNG)') { |v| options[:output] = v }
  opts.on('-c', '--color HEX', 'Цвет фона') { |v| options[:color] = v }
  opts.on('-t', '--threshold N', Integer, 'Порог (0-100)') { |v| options[:threshold] = v }
  opts.on('-v', '--verbose', 'Подробный вывод') { options[:verbose] = true }
  opts.on('-h', '--help', 'Справка') { puts opts; exit }
end.parse!

input = ARGV[0]
unless input
  puts "Ошибка: не указан входной файл."
  exit 1
end

unless File.exist?(input)
  puts "Ошибка: файл #{input} не найден"
  exit 1
end

output = options[:output] || input.sub(/\.[^.]+$/, '') + '_sticker.png'
threshold = options[:threshold] || 30
verbose = options[:verbose]

begin
  img = MiniMagick::Image.open(input)
  width = img.width
  height = img.height

  # Определяем цвет фона
  if options[:color]
    bg_color = MiniMagick::Image.read("xc:#{options[:color]}") { |c| c.format('png') }[0].get_pixels[0][0]
    bg_r, bg_g, bg_b = bg_color
  else
    # Автоопределение: средний цвет краёв
    pixels = []
    (0...width).each do |x|
      pixels << img.get_pixels[x][0]
      pixels << img.get_pixels[x][height-1]
    end
    (0...height).each do |y|
      pixels << img.get_pixels[0][y]
      pixels << img.get_pixels[width-1][y]
    end
    sr = sg = sb = 0
    pixels.each do |p|
      sr += p[0]; sg += p[1]; sb += p[2]
    end
    n = pixels.size
    bg_r = (sr / n).to_i
    bg_g = (sg / n).to_i
    bg_b = (sb / n).to_i
  end

  if verbose
    puts "Цвет фона: RGB(#{bg_r},#{bg_g},#{bg_b})"
  end

  max_dist = Math.sqrt(3 * 255 * 255)
  thresh = (threshold / 100.0) * max_dist

  # Создаём новое изображение с альфа-каналом
  # MiniMagick не поддерживает попиксельную установку альфа напрямую, поэтому используем `convert` через обработку.
  # Мы можем создать маску и применить её, но проще использовать подход с `-fuzz` и `-transparent`.
  # Однако это не даст точного контроля. Для демонстрации используем упрощённый подход с `-transparent-color`.
  # Но поскольку нам нужен точный порог, мы используем `-fuzz` и `-transparent`.
  # Альтернатива: использовать RMagick с попиксельным доступом, но это требует установки.

  # Вместо сложного попиксельного подхода, используем команду ImageMagick:
  # convert input.png -fuzz XX% -transparent "#RGB" output.png
  # Но это не позволяет указать цвет фона, отличный от одного цвета.
  # Для лучшего результата можно использовать несколько проходов.
  # Однако для демонстрации мы используем простой способ с `-transparent` и `-fuzz`.
  # Для более точного контроля в Ruby можно использовать RMagick.
  # Я предлагаю использовать упрощённый вариант: просто вызвать convert с -transparent.

  # Для корректной работы с RGB
  bg_hex = format('#%02x%02x%02x', bg_r, bg_g, bg_b)
  fuzz = threshold * 1.0
  # Создаём маску, но проще использовать -transparent
  # Делаем копию и применяем
  img2 = img.copy
  img2.combine_options do |c|
    c.fuzz("#{fuzz}%")
    c.transparent(bg_hex)
  end
  # Устанавливаем формат PNG и сохраняем
  img2.format('png')
  img2.write(output)
  img2.destroy

  if verbose
    puts "Фон удалён. Результат сохранён в #{output}"
  end
rescue => e
  puts "Ошибка: #{e.message}"
  exit 1
end
